File: mjlib_lua/base/hulib.lua
local table_mgr = require "table_mgr"
local M = {}

function M.check_7dui(hand_cards, waves)
    if #waves > 0 then return false end

    for _,c in ipairs(hand_cards) do
        if c % 2 ~= 0 then
            return false
        end
    end

    return true
end

function M.check_pengpeng()

end

function M.get_hu_info(hand_cards, waves, cur_card)
    local cards = {}
    for i,v in ipairs(hand_cards) do
        cards[i] = v
    end

    if cur_card then
        cards[cur_card] = cards[cur_card] + 1
    end

    local first_info = {
        eye = false,            -- 当前是否有将
    }

    -- 字牌（东西南北中发白)
    if not M.check_color(cards, 28, 34, first_info) then
        return false
    -- 万
    elseif not M.check_color_chi(cards, 1, 9, first_info) then
        return false
    -- 筒
    elseif not M.check_color_chi(cards, 10, 18, first_info) then
        return false
    -- 条
    elseif not M.check_color_chi(cards, 19, 27, first_info) then
        return false
    end
    return true
end

function M.check_color(cards, min, max, info)
    for i = min, max do
        local count = cards[i]

        if count == 1 or count == 4 then
            return false
        end

        if count == 2 then
            if info.eye then
                return false
            end

            info.eye = true
        end
    end

    return true
end

function M.check_color_chi(cards, min, max, info)
    local key = 0
    for i = min, max do
        local c = cards[i]
        if c > 0 then
            key = key * 10 + c
        elseif key > 0 then
            local eye = (key%3 == 2)
            if info.eye and eye then
                return false
            end

            if not table_mgr:check(key) then
                return false
            end
            key = 0
        end
    end

    return true
end

return M

File: mjlib_c++/base/hulib.cpp
#include <stdio.h>
#include <string.h>
#include "hulib.h"
#include "wave_table.h"
#include "wave_eye_table.h"

//#define LOG printf
#define LOG log

void log(char*, ...)
{
}

bool HuLib::get_hu_info(char* const hand_cards, Wave* const waves, char self_card, char other_card)
{
    char hand_cards_tmp[34];
    memcpy(hand_cards_tmp, hand_cards, 34);

    HuInfo info;

    if(!check_zi(hand_cards_tmp, info))
    {
        LOG("字牌检查失败 \n");
        return false;
    }

    if(!check_chi(hand_cards_tmp, info, 0))
    {
        LOG("万牌检查失败 \n");
        return false;
    }

    if(!check_chi(hand_cards_tmp, info, 9))
    {
        LOG("筒牌检查失败\n");
        return false;
    }

    if(!check_chi(hand_cards_tmp, info, 18))
    {
        LOG("条牌检查失败\n");
        return false;
    }

    return true;
}

bool HuLib::check_zi(char* const hand_cards, HuInfo& info)
{
    for(int i = 27; i<34; ++i)
    {
        int n = hand_cards[i];
        if(n == 1 || n == 4)
        {
            return false;
        }

        if(n == 2)
        {
            if(info.eye)
            {
                return false;
            }

            info.eye = true;
        }
    }

    return true;
}

bool HuLib::check_chi(char* const hand_cards, HuInfo& info, int min)
{
    char tbl[9];
    int n = 0;

    for(int i=min; i<min+9; ++i)
    {
        int c = hand_cards[i];
        if(c > 0)
        {
            tbl[n++] = c;
        }

        if(c == 0 || i == min + 9 - 1)
        {
            if(n==0) continue;

            if(!check_sub(tbl, n, info))
            {
                return false;
            }

            n = 0;
        }
    }

    return true;
}

bool HuLib::check_sub(char* const cards, int n, HuInfo& info)
{
    LOG("wave:  ");
    for(int i=0;i<n;++i)
    {
        LOG("%d ", cards[i]);
    }
    LOG("\n");

    int c = 0;
    int number = 0;
    for(int i=0;i<n;i++)
    {
        c = c + cards[i];
        number = number * 10 + cards[i];
    }

    int yu = c % 3;

    if(yu == 1)
    {
        return false;
    }
    else if(yu == 2)
    {
        if(info.eye) return false;

        info.eye = true;

        return check_wave_and_eye(number);
    }

    return check_wave(number);
}

bool HuLib::check_wave(int number)
{
    return WaveTable::getInstance()->check(number);
}

bool HuLib::check_wave_and_eye(int number)
{
    return WaveEyeTable::getInstance()->check(number);
}

bool HuLib::check_7dui(char* const cards)
{
    int c = 0;
    for(int i=0;i<34;++i)
    {
        if(cards[i] % 2 != 0) return false;
        c += cards[i];
    }

    if(c != 34) return false;

    return true;
}

bool HuLib::check_pengpeng(char* const cards, Wave* const waves, char self_card, char other_card)
{
    bool eye = false;

    for(int i=0;i<34;++i)
    {
        int c = cards[i];
        if(c == 1 || c == 4) return false;

        if(c == 2)
        {
            if(eye) return false;
            eye = true;
        }
    }

    // 检查吃碰杠的牌

    return true;
}

